druid-handler/src/java/org/apache/hadoop/hive/druid/DruidStorageHandler.java

```diff
@@ -62,28 +144,128 @@ public HiveMetaHook getMetaHook() {
   @Override
   public void preCreateTable(Table table) throws MetaException {
     // Do safety checks
-    if (!MetaStoreUtils.isExternalTable(table)) {
-      throw new MetaException("Table in Druid needs to be declared as EXTERNAL");
-    }
-    if (!StringUtils.isEmpty(table.getSd().getLocation())) {
-      throw new MetaException("LOCATION may not be specified for Druid");
-    }
+    /*
+    @TODO ASK Jesus if this is needed, i think we don't needed anymore
+    if (MetaStoreUtils.isExternalTable(table) && !StringUtils.isEmpty(table.getSd().getLocation())) {
+      throw new MetaException("LOCATION may not be specified for Druid existing sources");
+    }*/
     if (table.getPartitionKeysSize() != 0) {
       throw new MetaException("PARTITIONED BY may not be specified for Druid");
     }
     if (table.getSd().getBucketColsSize() != 0) {
       throw new MetaException("CLUSTERED BY may not be specified for Druid");
     }
+    String dataSourceName = Preconditions
+            .checkNotNull(table.getParameters().get(Constants.DRUID_DATA_SOURCE),
+                    "WTF dataSource name is null !"
+            );
+    try {
```

Comment:
That looks good; if it is external, you could even verify here if the datasource exists (it should), as you already retrieve the existing datasources. But maybe it is not needed, as we might have checked it before. It is something that we might double-check while testing.